rotkehlchen/chain/bitcoin/__init__.py

```diff
@@ -13,65 +15,130 @@ def _have_bc1_accounts(accounts: List[BTCAddress]) -> bool:
+                deserialize_fval(
+                    value=entry['final_balance'],
+                    name='final_balance',
+                    location='blockchain.info api',
+                ),
+            )
+    return balances
+
+
+def get_bitcoin_addresses_balances(
+        accounts: List[BTCAddress],
+) -> Dict[BTCAddress, FVal]:
+    """Queries bitcoin balance APIs for the balances of accounts
+
+    May raise:
+    - RemoteError couldn't query any of the bitcoin balance APIs
+    """
+    if _have_bc1_accounts(accounts) is True:
+        api_callbacks = {
+            'blockstream.info': _query_blockstream_info,
+            'mempool.space': _query_mempool_space,
+        }
+    else:
+        api_callbacks = {
+            'blockchain.info': _query_blockchain_info,
+            'blockstream.info': _query_blockstream_info,
+            'mempool.space': _query_mempool_space,
+        }
+    errors: Dict[str, str] = {}
+    for api_name, callback in api_callbacks.items():
+        try:
+            balances = callback(accounts)
+        except (
             requests.exceptions.RequestException,
             UnableToDecryptRemoteData,
             requests.exceptions.Timeout,
-    ) as e:
-        raise RemoteError(f'bitcoin external API request for balances failed due to {str(e)}') from e  # noqa: E501
-    except KeyError as e:
-        raise RemoteError(
-            f'Malformed response when querying bitcoin blockchain via {source}.'
-            f'Did not find key {e}',
-        ) from e
+            RemoteError,
+            DeserializationError,
+        ) as e:
+            errors[api_name] = str(e)
+            continue
+        except KeyError as e:
+            errors[api_name] = f'Got unexpected response from {api_name}. Couldn\'t find key {str(e)}'  # noqa: E501
+        else:
+            return balances
 
-    return balances
+    serialized_errors = ', '.join(f'{source} error is: "{error}"' for (source, error) in errors.items())  # noqa: E501
+    raise RemoteError(f'Bitcoin external API request for balances failed. {serialized_errors}')  # noqa: E501
 
 
 def _check_blockstream_for_transactions(
         accounts: List[BTCAddress],
 ) -> Dict[BTCAddress, Tuple[bool, FVal]]:
-    """May raise connection errors or KeyError"""
+    """May raise:
+    - RemoteError if couldn't query
+    - KeyError if response structure differs from the expected one
+    - DeserializationError if response values differ from the expected
+    """
     have_transactions = {}
     for account in accounts:
         url = f'https://blockstream.info/api/address/{account}'
         response_data = request_get_dict(url=url, handle_429=True, backoff_in_seconds=4)
         stats = response_data['chain_stats']
-        balance = satoshis_to_btc(int(stats['funded_txo_sum']) - int(stats['spent_txo_sum']))
+        funded_txo_sum = deserialize_fval(value=stats['funded_txo_sum'], name='funded_txo_sum', location='blockstream api')  # noqa: E501
```

Comment:
If this code is also the same as `_query_mempool_space` and `_query_blockstream_info` use the same abstract function you will use there, here too.

Reply:
How do see this done? I don't think it will make sense to do since this function not only calculates `balance` but also accesses `stats`, so our abstract function should return `stats` object which will be unused in other querying functions